src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php

```diff
@@ -171,4 +182,20 @@ public function findOneByCode(string $code): ?ProductInterface
             ->getOneOrNullResult()
         ;
     }
+
+    private function reproduceTaxonomyTree(TaxonInterface $taxon, QueryBuilder $queryBuilder): void
+    {
+        if (!$taxon->hasChildren()) {
+            $parameter = 'taxonChild' . $taxon->getCode();
+
+            $queryBuilder->orWhere('productTaxon.taxon = :' . $parameter);
+            $queryBuilder->setParameter($parameter, $taxon);
+
+            return;
+        }
+
+        foreach ($taxon->getChildren() as $taxonChild) {
+            $this->reproduceTaxonomyTree($taxonChild, $queryBuilder);
```

Comment:
I'm afraid that it would do tons of queries to the database just to get that taxonomy tree (and the final SQL would be gargantuan).

Reply:
Have you tried using `productTaxon.taxon.left` and `productTaxon.taxon.right` for filtering? We use nested set for taxons: https://en.wikipedia.org/wiki/Nested_set_model